File: StriverA2Z/C++/3_ArraysQuestion/3.3_Hard/subarray_with_xor_k.cpp
#include<bits/stdc++.h>

using namespace std;

void subarrayXorK(int arr[], int n, int target) {
    map<int, int> mpp;
    int xr = 0;
    mpp[xr]++;
    int cnt = 0;
    for (int i = 0; i < n; i++) {
        xr ^= arr[i];
        int x = xr ^ target;
        cnt += mpp[x];
        mpp[xr]++;
    }
    cout << cnt;
}

int main() {
    int arr[] = {4, 2, 2, 6, 4};
    subarrayXorK(arr, sizeof(arr)/ sizeof(arr[0]), 6);
    return 0;
}

File: StriverA2Z/Java/Arrays/Hard/SubarrayWithXorK.java
import java.util.HashMap;

public class SubarrayWithXorK {
    public static void main(String[] args) {
        int[] arr = {5, 6, 7, 8, 9};
        subarrayXorK(arr, 5);
    }

    static void subarrayXorK(int[] arr, int target) {
        int xr = 0;
        HashMap<Integer, Integer> mpp = new HashMap<>();
        mpp.put(0, 1);
        int cnt = 0;
        for (int j : arr) {
            xr ^= j;
            int x = xr ^ target;
            cnt += mpp.getOrDefault(x, 0);
            mpp.put(xr, mpp.getOrDefault(xr, 0) + 1);
        }
        System.out.println(cnt);
    }
}
